### ISSUE TITLE
502 Bad Gateway nginx/1.15.8

### ISSUE THREAD

**Issue by Flerex** [laravel/valet#717]
I'm getting a 502 Bad Gateway nginx/1.15.8 error when trying to enter a .test site. This is on a **completely fresh Mojave installation** where I just installed with Homebrew `php` (7.3.1), `nginx` (1.15.8)  (among others) and `Valet` (2.1.6) through composer. I haven't modified any default configuration of any of the aforementioned packages. I have run `valet install` and `valet park` in the folder where my project is at.

Any idea why this is happening?

**Comment by drbyte** [comment]
Hmmm ... I've had no issues with 7.3 on a fresh Mojave install.

Maybe I'll wipe this new machine and test again.

**Comment by Flerex** [comment]
@mattstauffer How am I supposed to downgrade? I tried `brew uninstall php && brew install php@7.2`,   made sure it's been added to the `$PATH` and when I try to do `valet install` I get the following error:

> Homebrew PHP appears not to be linked. 

@drbyte You had no issues? That's weird. What can I be doing wrong?

**Comment by Flerex** [comment]
@mattdfloyd 7.2 installed, same thing happening.

**Comment by drbyte** [comment]
As I understand it, Bad Gateway means Nginx isn't finding the PHP-FPM service to proxy (gateway) the request through.

Common causes:
- PHP isn't running
- the valet config isn't finding the `valet.sock` socket, which is created by the PHP process when it starts
- the valet nginx configs are pointing to the wrong php version

**Comment by florianbouvot** [comment]
Did you try this 

´´´
brew services start php
´´´

If it didn’t work, try to reinstall php from source

**Comment by aurelianoa** [comment]
I know this is closed but i had the same issue and i was because some how brew had running an older php version so:
`sudo brew services cleanup`
and after i restart it worked. i have Mac OSX High Sierra btw

**Comment by iqigmbh** [comment]
A simple restart of the brew services with sudo in front and an adjacent restart of valet did the job for me:

`sudo brew services restart nginx && sudo brew services restart php`
`valet restart`

**Comment by DrewRoberts** [comment]
I ran into this issue as well. After doing the commands shared in the other responses, I also had to change back to php 7.2 instead of 7.3 and here's that command:

`valet use php@7.2`

More about it here: [Laravel Valet PHP Versions](https://laravel.com/docs/5.8/valet#installation)

**Comment by araeuchle** [comment]
Happened to me as well after i upgraded my macos with brew update && brew upgrade

valet use php@7.2 did the trick for. thanks @drewroberts, you saved me lots of hours :)

**Comment by drbyte** [comment]
I ran `brew upgrade && brew update && brew cleanup` recently and it upgraded me from 7.3.5 to 7.3.6 and is working fine. 
It's been awhile since I bothered with PHP 7.2, so maybe your glitch was in moving from 7.2 to 7.3. 

The Bad Gateway message means Nginx can't find a running PHP instance, as determined by having PHP-FPM properly configured to create the valet.sock file so Nginx can use that to send PHP request through. The default valet configs for PHP-FPM work fine on my machine.
If yours aren't working, you can delete your /usr/local/etc/php directory (backups first of course) and then re-install 7.3.

**Comment by skakrecha** [comment]
> Happened to me as well after i upgraded my macos with brew update && brew upgrade
> 
> valet use php@7.2 did the trick for me. thanks @drewroberts, you saved me lots of hours :)

I cant use the "valet use" it says command use not defined

**Comment by drbyte** [comment]
> I cant use the "valet use" it says command use not defined

That suggests you're using an old version of Valet. I believe the `use` command was added in 2.2.0

**Comment by josermorales** [comment]
Fixed it by doing
```
brew services start --all
```

**Comment by Ahmed-Aboud** [comment]
a really weird reason that solved it was a missing default value just as i passed it the POST method worked

**Comment by ElectroBuddha** [comment]
Tried all of the suggestions, fixed only after I updated valet to the latest version (v2.3.3):
`composer global update`
`valet install`

**Comment by metallurgical** [comment]
Both @ElectroBuddha and @drewroberts did the tricks

**Comment by rcordobar19** [comment]
> Started having this issue after I updated php to v7.3.
> Tried all of the suggestions, fixed only after I updated valet to the latest version (v2.3.3):
> `composer global update`
> `valet install`

Thank you, it worked

**Comment by mugisham** [comment]
valet unistall
valet install 
valet park

these 3 steps worked for me

**Comment by abhisheksatre** [comment]
> Started having this issue after I updated php to v7.3.
> Tried all of the suggestions, fixed only after I updated valet to the latest version (v2.3.3):
> `composer global update`
> `valet install`

Thank you. it worked

**Comment by OliverGrimsley** [comment]
Seems every time I update/upgrade brew I have this issue (nginx 1.17.7).  Had it again when going to PHP 7.4.  Here is what I had to do, not sure if the last thing I did would have worked first, so, the last thing I did was:

`valet use php@7.4`

Before that, I completely eradicated PHP (note, at least on Catalina, you cannot use sudo in front of brew, or I do not know how):
```
brew doctor
brew services cleanup
brew list | grep php
brew uninstall --force php@[INSERT VERSIONS]
```
Re-ran `brew doctor` and `brew services cleanup` and `brew cleanup` and had to manually `sudo rm -R [directories]` where directories were in `usr\local\etc\` then

```brew install php@7.4
valet restart
valet parked
valet use php@7.4
```

Brew is such a PITA.  I had to `chown` a number of directories to fix permissions, `rm` stuff etc.

**Comment by drbyte** [comment]
FYI - "502 Bad Gateway" from Nginx means the PHP proxy is not running, or not found. (Nginx proxies PHP calls to php-fpm by means of the directives in the nginx site conf files.)

So, if you've updated your PHP to a new non-minor version (ie: 7.2 up to 7.3, or 7.4 down to 7.3) by just running a `brew upgrade` command, then you'll need to also run `valet install` to configure that PHP version's configs for Valet. Otherwise your newly-installed PHP version's configs won't know anything about Valet.

If you upgrade PHP or switch to new PHP version via the `valet use php@x.x` command  then Valet will also do the necessary config updates as part of that switch.

My usual workflow (and I normally only use the latest PHP version anyway) is simply the following on a weekly basis:
- composer global update
- brew upgrade
and if homebrew gave me a new major PHP version (ie: 7.4 when I previously had 7.3) then I also run:
- valet install

**Comment by liyansasongko** [comment]
> Did you try this
> 
> ```
> brew services start php
> ```
> 
> If it didn’t work, try to reinstall php from source
> 
> ```
> brew uninstall php
> brew install php --build-from-source
> valet install
> ```

it worked

**Comment by liyansasongko** [comment]
@florianbouvot

**Comment by jordanade** [comment]
I'm getting this 502 error—but only when a 500 error would have otherwise been produced. (So no Ignition error page, which is sad.) None of these suggestions are working for me.

**Comment by drbyte** [comment]
> I'm getting this 502 error—but only when a 500 error would have otherwise been produced. (So no Ignition error page, which is sad.) None of these suggestions are working for me.

If a 500 error is occurring behind the scenes, then you'll need to go back to old-school troubleshooting by consulting the logs. You mentioned Ignition which suggests you're working with a Laravel app, which means your logs will be in Laravel's usual log location.

**Comment by jordanade** [comment]
> > I'm getting this 502 error—but only when a 500 error would have otherwise been produced. (So no Ignition error page, which is sad.) None of these suggestions are working for me.
> 
> If a 500 error is occurring behind the scenes, then you'll need to go back to old-school troubleshooting by consulting the logs. You mentioned Ignition which suggests you're working with a Laravel app, which means your logs will be in Laravel's usual log location.

Thanks drbyte. Yes I'm using the latest version of Laravel and Valet. It's just that this is previously unseen behavior with standard PHP errors, and happened after I did some PHP/Homebrew updating. Routes without errors are working fine.

**Comment by saiftheboss7** [comment]
Thanks @rcordobar19 That fixed it.

**Comment by zeshan77** [comment]
The following commands worked for me.

`valet use php@7.2`
`composer global update`

**Comment by andresams** [comment]
Upgrading to php 7.4 worked for me.

`valet use 7.4`

**Comment by kirkbushell** [comment]
None of these solutions worked for me, and am just about at my wit's end. I want so bad to use valet but the nginx/php-fpm issue is quite clearly a main problem that needs to be fixed. I understand that php-fpm isn't running or getting setup correctly, but the feedback the commands are giving me is not enough to figure out the why, and as mentioned - no solutions here have worked. Any ideas how to debug/resolve?

**Comment by drbyte** [comment]
Hey @kirkbushell , do you mind opening a new issue, and posting the output of `valet diagnose` as part of it? That'll help with ruling things out as well as potentially identifying related factors that might not be obvious enough yet.

**Comment by kirkbushell** [comment]
@drbyte actually it's funny, right as I was about to give up, I found an article that helped me resolve it - and it all came down to path export in zshrc.

**Comment by drbyte** [comment]
@kirkbushell I'm glad you got it sorted out. If you think the article points to something that's not rather unique to your situation but rather is likely to be common for others to benefit from, feel free to share the link when you have a minute.

Cheers!

**Comment by kirkbushell** [comment]
For me, it was this article that set me straight: 

https://programmer.help/blogs/5e2800f291b7a.html

Specifically, the point about paths in zshrc.

**Comment by rcordobar19-mbm** [comment]
FYI,
I updated to PHP 7.49 and tried using it on valet. Got:

In PhpFpm.php line 212:
Valet doesn't support PHP version: php@7.49 (try something like 'php@7.3' instead).

So is something to take into consideration.

**Comment by rcordobar19** [comment]
FYI, 
Just updated to PHP 7.49 and tried using valet. Got:
In PhpFpm.php line 212:
Valet doesn't support PHP version: php@7.49 (try something like 'php@7.3' instead)

So, is something to take into consideration.

**Comment by patrickoliveras** [comment]
In my case php was unlinked but already installed so running `brew link php@7.3 --force` and restarting valet did the trick

**Comment by mantey-github** [comment]
> I ran into this issue as well. After doing the commands shared in the other responses, I also had to change back to php 7.2 instead of 7.3 and here's that command:
> 
> `valet use php@7.2`
> 
> More about it here: [Laravel Valet PHP Versions](https://laravel.com/docs/5.8/valet#installation)

Thanks. valet use php@7.4 also worked

**Comment by johnvoncolln** [comment]
> I'm getting this 502 error—but only when a 500 error would have otherwise been produced. (So no Ignition error page, which is sad.) None of these suggestions are working for me.

@jordanade did you ever find a solution? Having the same problem after upgrading to bigSur...

**Comment by jordanade** [comment]
> > I'm getting this 502 error—but only when a 500 error would have otherwise been produced. (So no Ignition error page, which is sad.) None of these suggestions are working for me.
> 
> @jordanade did you ever find a solution? Having the same problem after upgrading to bigSur...

I can't quite remember how I fixed it now. I think maybe it was due to multiple PHP installations running under `brew services`, you may want to check that.

**Comment by johnvoncolln** [comment]
> > > I'm getting this 502 error—but only when a 500 error would have otherwise been produced. (So no Ignition error page, which is sad.) None of these suggestions are working for me.
> > 
> > 
> > @jordanade did you ever find a solution? Having the same problem after upgrading to bigSur...
> 
> I can't quite remember how I fixed it now. I think maybe it was due to multiple PHP installations running under `brew services`, you may want to check that.

Thanks - after like four days and maybe a dozen or so reinstalls of Valet, I deleted the project that was giving me trouble, recloned it and everything worked.  Not sure what the issue was, but maybe this will help someone else.

**Comment by mertasan** [comment]
`brew services stop php`
`brew services start php`

```cli
$ brew services start php
Bootstrap failed: 5: Input/output error
Error: Failure while executing; `/bin/launchctl bootstrap gui/501 /Users/***/Library/LaunchAgents/homebrew.mxcl.php.plist` exited with 5.
$ brew services stop php
Stopping `php`... (might take a while)
==> Successfully stopped `php` (label: homebrew.mxcl.php)
$ brew services start php
==> Successfully started `php` (label: homebrew.mxcl.php)
$ valet restart
Restarting dnsmasq...
Restarting php...
Restarting nginx...
Valet services have been restarted.
```

Fixed.

**Comment by usamamuneerchaudhary** [comment]
`valet use php@8.1 --force` This worked for me.

**Comment by moedayraki** [comment]
Reinstalling valet after installing php@8.1 worked for me `valet install`

**Comment by AlexShevchenko1985** [comment]
> valet unistall valet install valet park
> 
> these 3 steps worked for me

this helped me!!

**Comment by VJSriram** [comment]
![image](https://user-images.githubusercontent.com/23410987/170581300-ae333233-510d-4214-9064-93a3bd20ef1c.png)

**Comment by rtconner** [comment]
Is everyone else who is consistently having this issue using a Macbook and jumping from wifi to wifi all the time?

**Comment by SenaJp** [comment]
I'm using the PHP version isolation functionality, it only worked after I removed the isolation and put the version that matches the project.

**Comment by nguyenquocdaile** [comment]
> Did you try this
> 
> ```
> brew services start php
> ```
> 
> If it didn’t work, try to reinstall php from source
> 
> ```
> brew uninstall php
> brew install php --build-from-source
> valet install
> ```

It's worked for me. I guess the error occurred after I had installed phpfpm

**Comment by Mattgh91** [comment]
I tried a lot of these with no luck, the fix for me was to re-run 
`valet isolate php@8.3`
on the affected sites.
